sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java

```diff
@@ -211,7 +211,10 @@ public int compare(Object baseObj1, long baseOff1, Object baseObj2, long baseOff
       // TODO: Why are the sizes -1?
       row1.pointTo(baseObj1, baseOff1, -1);
       row2.pointTo(baseObj2, baseOff2, -1);
-      return ordering.compare(row1, row2);
+      int comparison = ordering.compare(row1, row2);
+      row1.pointTo(null, 0L, -1);
+      row2.pointTo(null, 0L, -1);
```

Comment:
maybe it's worth having a clone method on the comparator and making sure we clone the comparator before passing it to anything?